codec/src/main/java/io/netty/handler/codec/DatagramPacketEncoder.java

```diff
@@ -0,0 +1,123 @@
+ * pipeline.addLast("udpEncoder", new {@link DatagramPacketEncoder}(new {@link ProtobufEncoder}(...));
+ * </code></pre>
+ *
+ * Note: As UDP packets are out-of-order, you should make sure the encoded message size are not greater than
+ * {@code maxPacketSize} which is the max safe packet size in your particular network path which guarantees
+ * no packet fragmentation.
+ *
+ * @param <M> the type of message to be encoded
+ */
+public class DatagramPacketEncoder<M> extends MessageToMessageEncoder<AddressedEnvelope<M, InetSocketAddress>> {
+
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(DatagramPacketEncoder.class);
+
+    private final int maxPacketSize;
+
+    private final MessageToMessageEncoder<? super M> encoder;
+
+    /**
+     * Create an encoder that encodes the content in {@link AddressedEnvelope} to {@link DatagramPacket} using
+     * the specified message encoder.
+     *
+     * @param encoder the specified message encoder
+     */
+    public DatagramPacketEncoder(MessageToMessageEncoder<? super M> encoder) {
+        this(encoder, 512);
+    }
+
+    /**
+     * Create an encoder that encodes the content in {@link AddressedEnvelope} to {@link DatagramPacket} using
+     * the specified message encoder.
+     *
+     * @param encoder the specified message encoder
+     * @param maxPacketSize the max safe packet size in your particular network path which guarantees
+     *                      no packet fragmentation
+     */
+    public DatagramPacketEncoder(MessageToMessageEncoder<? super M> encoder, int maxPacketSize) {
+        this.encoder = encoder;
+        this.maxPacketSize = maxPacketSize;
+    }
+
+    @Override
+    public boolean acceptOutboundMessage(Object msg) throws Exception {
+        if (super.acceptOutboundMessage(msg)) {
+            AddressedEnvelope envelope = (AddressedEnvelope) msg;
+            return encoder.acceptOutboundMessage(envelope.content())
+                    && envelope.sender() instanceof InetSocketAddress
+                    && envelope.recipient() instanceof InetSocketAddress;
+        }
+        return false;
+    }
+
+    @Override
+    protected void encode(
+            ChannelHandlerContext ctx, AddressedEnvelope<M, InetSocketAddress> msg, List<Object> out) throws Exception {
+        RecyclableArrayList buffers = null;
+        try {
+            buffers = RecyclableArrayList.newInstance();
+            encoder.encode(ctx, msg.content(), buffers);
+            if (buffers.size() != 1) {
+                throw new EncoderException(
+                        StringUtil.simpleClassName(encoder) + " must produce only one message.");
+            }
+            Object content = buffers.get(0);
+            if (content instanceof ByteBuf) {
+                ByteBuf buf = (ByteBuf) content;
+                if (buf.readableBytes() > maxPacketSize) {
+                    logger.warn("The datagram packet size {} is large than {}. " +
```

Comment:
Do we even need `maxPacketSize` at all? There's no way to guarantee its the actual maximum length of a datagram packet enforced by the network.